MDEV-36479 draft on checking a bit more strictness

grooverdan opened the pull request:
- [x] *The Jira issue number for this PR is: MDEV-______ (not yet)*

<!--
An amazing description should answer some questions like:
1. What problem is the patch trying to solve?
2. If some output changed that is not visible in a test case, what was it looking like before the change and how it's looking with this patch applied?
3. Do you think this patch might introduce side-effects in other parts of the server?
-->
## Description

Hiding over uninitialised variable seems to be from detection bugs in old compilers. More imporant, its leading to current undefined behaviour. Lets see if that's still the case.

e.g.:
```
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /home/dan/repos/mariadb-server-11.8/mysys/mf_iocache.c:1793:39 
=================================================================
==2772984==ERROR: AddressSanitizer: use-after-poison on address 0x52d0003370b0 at pc 0x000000d6cc9c bp 0x7fff321294f0 sp 0x7fff321294e8
READ of size 8 at 0x52d0003370b0 thread T0
    #0 0x000000d6cc9b in multi_update::num_found() const /home/dan/repos/mariadb-server-11.8/sql/sql_class.h:7665:45
    #1 0x000000d6cc9b in MYSQL_DML_DONE(THD*, int) /home/dan/repos/mariadb-server-11.8/sql/sql_select.cc:34329:5
    #2 0x000000d6af0d in Sql_cmd_dml::execute(THD*) /home/dan/repos/mariadb-server-11.8/sql/sql_select.cc:34504:3
    #3 0x000000b0f5ef in mysql_execute_command(THD*, bool) /home/dan/repos/mariadb-server-11.8/sql/sql_parse.cc:4428:27
    #4 0x000000af38fc in mysql_parse(THD*, char*, unsigned int, Parser_state*) /home/dan/repos/mariadb-server-11.8/sql/sql_parse.cc:7915:18
    #5 0x000000af15d7 in bootstrap(st_mysql_file*) /home/dan/repos/mariadb-server-11.8/sql/sql_parse.cc:1090:5
    #6 0x000000603dd5 in mysqld_main(int, char**) /home/dan/repos/mariadb-server-11.8/sql/mysqld.cc:6091:26
    #7 0x7f3142a10247 in __libc_start_call_main (/lib64/libc.so.6+0x3247) (BuildId: f83d43b9b4b0ed5c2bd0a1613bf33e08ee054c93)
    #8 0x7f3142a1030a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x330a) (BuildId: f83d43b9b4b0ed5c2bd0a1613bf33e08ee054c93)
    #9 0x00000051a7b4 in _start (/home/dan/repos/build-mariadb-server-11.8/sql/mariadbd+0x51a7b4) (BuildId: 365124ed2d3a4abc252a508657b91cd007b3c1d6)
```

Put nonnull on the character set as its rather pointless to to do comparisons with null pointers and the approprate logic to take was undoubtedly at a higher level.
```
/home/dan/repos/mariadb-server-11.8/include/m_ctype.h:1949:33: runtime error: null pointer passed as argument 1, which is declared to never be null
/home/dan/repos/mariadb-server-11.8/include/m_ctype.h:1035:65: note: nonnull attribute specified here
    #0 0x000000700974 in Well_formed_prefix_status::Well_formed_prefix_status(charset_info_st const*, char const*, char const*, unsigned long) /home/dan/repos/mariadb-server-11.8/include/m_ctype.h:1949:9
    #1 0x000000b476dc in Well_formed_prefix::Well_formed_prefix(charset_info_st const*, char const*, unsigned long, unsigned long) /home/dan/repos/mariadb-server-11.8/include/m_ctype.h:1963:5
    #2 0x000000b476dc in check_string_char_length(st_mysql_const_lex_string const*, unsigned int, unsigned long, charset_info_st const*, bool) /home/dan/repos/mariadb-server-11.8/sql/sql_parse.cc:10118:22
    #3 0x000000b34aea in check_ident_length(st_mysql_const_lex_string const*) /home/dan/repos/mariadb-server-11.8/sql/sql_parse.cc:10136:7
    #4 0x000000e927e4 in mysql_prepare_create_table_finalize(THD*, HA_CREATE_INFO*, Alter_info*, unsigned int*, handler*, st_key**, unsigned int*, int) /home/dan/repos/mariadb-server-11.8/sql/sql_table.cc:3452:9
    #5 0x000000e8e121 in mysql_create_frm_image(THD*, HA_CREATE_INFO*, Alter_info*, int, st_key**, unsigned int*, st_mysql_const_unsigned_lex_string*) /home/dan/repos/mariadb-server-11.8/sql/sql_table.cc:4544:7
    #6 0x000000e9d154 in create_table_impl(THD*, st_ddl_log_state*, st_ddl_log_state*, Lex_ident_db const&, Lex_ident_table const&, st_mysql_const_lex_string const&, st_mysql_const_lex_string const&, st_mysql_const_lex_string const&, DDL_options_st, HA_CREATE_INFO*, Alter_info*, int, bool*, st_key**, unsigned int*, st_mysql_const_unsigned_lex_string*) /home/dan/repos/mariadb-server-11.8/sql/sql_table.cc:4866:11
    #7 0x000000e9b4b0 in mysql_create_table_no_lock(THD*, st_ddl_log_state*, st_ddl_log_state*, Table_specification_st*, Alter_info*, bool*, int, TABLE_LIST*) /home/dan/repos/mariadb-server-11.8/sql/sql_table.cc:4991:8
    #8 0x000000ef2c3a in mysql_create_table(THD*, TABLE_LIST*, Table_specification_st*, Alter_info*) /home/dan/repos/mariadb-server-11.8/sql/sql_table.cc:5233:7
    #9 0x000000eed278 in Sql_cmd_create_table_like::execute(THD*) /home/dan/repos/mariadb-server-11.8/sql/sql_table.cc:13624:12
    #10 0x000000b0efea in mysql_execute_command(THD*, bool) /home/dan/repos/mariadb-server-11.8/sql/sql_parse.cc:5886:26
    #11 0x000000af38fc in mysql_parse(THD*, char*, unsigned int, Parser_state*) /home/dan/repos/mariadb-server-11.8/sql/sql_parse.cc:7915:18
    #12 0x000000af15d7 in bootstrap(st_mysql_file*) /home/dan/repos/mariadb-server-11.8/sql/sql_parse.cc:1090:5
    #13 0x000000603dd5 in mysqld_main(int, char**) /home/dan/repos/mariadb-server-11.8/sql/mysqld.cc:6091:26
    #14 0x7f3142a10247 in __libc_start_call_main (/lib64/libc.so.6+0x3247) (BuildId: f83d43b9b4b0ed5c2bd0a1613bf33e08ee054c93)
    #15 0x7f3142a1030a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x330a) (BuildId: f83d43b9b4b0ed5c2bd0a1613bf33e08ee054c93)
    #16 0x00000051a7b4 in _start (/home/dan/repos/build-mariadb-server-11.8/sql/mariadbd+0x51a7b4) (BuildId: 365124ed2d3a4abc252a508657b91cd007b3c1d6)
```

Where are the errors?
```
 mysql-test/mtr --mem  --boot-gdb='b __asan_report_error;b __ubsan::ScopedReport::~ScopedReport;r' main.select
```

And remove `boot-` one those are found/ignored.

## Release Notes

draft only - not for release

## How can this PR be tested?

existing mtr tests.

Though  `-DWITH_ASAN=ON -DWITH_UBSAN=ON -DWITH_ASAN_SCOPE=ON -DPLUGIN_PERFSCHEMA=NO -DPLUGIN_MROONGA=NO -DCMAKE_BUILD_TYPE=Debug -DWITH_DBUG_TRACE=OFF` under modern clang highlight calls.
 
TODO: modify the automated test suite to verify that the PR causes MariaDB to behave as intended.
Consult the documentation on ["Writing good test cases"](https://mariadb.org/get-involved/getting-started-for-developers/writing-good-test-cases-mariadb-server).
<!--
In many cases, this will be as simple as modifying one `.test` and one `.result` file in the `mysql-test/` subdirectory.
Without automated tests, future regressions in the expected behavior can't be automatically detected and verified.
-->

If the changes are not amenable to automated testing, please explain why not and carefully describe how to test manually.

<!--
Tick one of the following boxes [x] to help us understand if the base branch for the PR is correct.
-->
## Basing the PR against the correct MariaDB version
- [ ] *This is a new feature or a refactoring, and the PR is based against the `main` branch.*
- [ ] *This is a bug fix, and the PR is based against the earliest maintained branch in which the bug can be reproduced.*

<!--
  All code merged into the MariaDB codebase must meet a quality standard and codying style.
  Maintainers are happy to point out inconsistencies but in order to speed up the review and merge process we ask you to check the CODING standards.
-->
## PR quality check
- [ ] I checked the [CODING_STANDARDS.md](https://github.com/MariaDB/server/blob/-/CODING_STANDARDS.md) file and my PR conforms to this where appropriate.
- [ ] For any trivial modifications to the PR, I am ok with the reviewer making the changes themselves.

grooverdan committed:
MDEV-36479 Passing null pointer to low level character set functions

results in undefined behaviour. Let's mark NONNULL as an attribute
to discover these problems at the source.

inspired by MySQL header file by Jan Kneschke @Oracle.

ref: MySQL Bug#32117977
     commit: d59e58d595b8e43026347f9438a81bc1d0ebb8c3
     RB: 25407
     file: router/src/harness/include/mysql/harness/stdx/attribute.h

FooBarrior commented:
@grooverdan maybe instead of gcc-style __attribute__, let's add MY_NONNULL macro to `my_compiler.h`?

grooverdan committed:
MDEV-40176: field_charset()->charpos(blob..) called with NULL

Problem:
  When `my_charpos_mb()` is called by `Field_blob::get_key_image_itRAW`
  with a start/end both being NULL. Because of this the blob_length
  must have been 0.

Fix:
  Rather than relying on character set functions to calculate the
  storage of nothing, bypass the calculation as the charpos() isn't
  going to return a value less than 0 (the blob_length).

grooverdan committed:
MDEV-40234 String_copier::well_formed_copy inefficient for appending 0 bytes

For all the copying in Field*str::store, lets make a quick
path when we're copy 0 bytes in a compatible character set.

Noting Bar wants this check at a higher level.

grooverdan committed:
MDEV-40233 HEX(emtpy ascii string) could be quicker

Item_func_hex::val_str_ascii_from_val_str(str) when
str->ptr() is null, take a long path through returning
an empty string.

Replace this path with a quick return of an make_empty_result().

Also defer thd initialization until after a potential
null return. current_thd as implemented isn't cheap.

grooverdan committed:
spider fixes - no appending a nullptr

grooverdan committed:
compare nullptr and 0

test: rpl.rpl_ignore_grant

hack work around for:

(rr) p *res1
$21 = {<Charset> = {m_charset = 0x7988d20 <my_charset_utf8mb4_bin>}, <Binary_string> = {<Sql_alloc> = {<No data fields>}, Ptr = 0x0, str_length = 0,
    Alloced_length = 0, extra_alloc = 0, alloced = false, thread_specific = false}, <No data fields>}
(rr) p res2
$22 = (String *) 0x7e5d5af06cb0
(rr) p *res2
$23 = {<Charset> = {m_charset = 0x7988d20 <my_charset_utf8mb4_bin>}, <Binary_string> = {<Sql_alloc> = {<No data fields>}, Ptr = 0x7b7d5ae66a28 "",
    str_length = 0, Alloced_length = 0, extra_alloc = 0, alloced = true, thread_specific = false}, <No data fields>}
(rr) bt
    at /home/dan/repos/mariadb-server-11.8/sql/sql_string.cc:846
    at /home/dan/repos/mariadb-server-11.8/sql/sql_type.cc:7748
    at /home/dan/repos/mariadb-server-11.8/sql/sql_type.h:6003
    at /home/dan/repos/mariadb-server-11.8/sql/item.h:1268
    at /home/dan/repos/mariadb-server-11.8/sql/protocol.cc:1359
    items=@0x7e1d5b2a47f8: {<base_list> = {<Sql_alloc> = {<No data fields>}, first = 0x7e1d5b2a4e18, last = 0x7e1d5b2a4e18, elements = 1}, <No data fields>})
    at /home/dan/repos/mariadb-server-11.8/sql/sql_class.cc:3306
    items=@0x7e1d5b2a47f8: {<base_list> = {<Sql_alloc> = {<No data fields>}, first = 0x7e1d5b2a4e18, last = 0x7e1d5b2a4e18, elements = 1}, <No data fields>},
    u=0x7e0d5b024768, sent=0) at /home/dan/repos/mariadb-server-11.8/sql/sql_class.h:6380
    at /home/dan/repos/mariadb-server-11.8/sql/sql_select.cc:25757
    at /home/dan/repos/mariadb-server-11.8/sql/sql_select.cc:24644
    at /home/dan/repos/mariadb-server-11.8/sql/sql_select.cc:24448
    fields=@0x7e1d5b2a47f8: {<base_list> = {<Sql_alloc> = {<No data fields>}, first = 0x7e1d5b2a4e18, last = 0x7e1d5b2a4e18, elements = 1}, <No data fields>},
    conds=0x7e1d5b2a5840, og_num=0, order=0x0, group=0x0, having=0x0, proc_param=0x0, select_options=2164525824, result=0x7e5d5af00168, unit=0x7e0d5b024768,
    select_lex=0x7e1d5b2a4540) at /home/dan/repos/mariadb-server-11.8/sql/sql_select.cc:5426
    at /home/dan/repos/mariadb-server-11.8/sql/sql_select.cc:634
    at /home/dan/repos/mariadb-server-11.8/sql/sql_parse.cc:4014
    rawbuf=0x7e1d5b2a4438 "select password<>_binary'' from mysql.user where user=_binary'rpl_ignore_grant'", length=79, parser_state=0x734d38c247b0)
    at /home/dan/repos/mariadb-server-11.8/sql/sql_parse.cc:7953

grooverdan committed:
hack perfschema.start_server_zero_digest_sql_length

    at /home/dan/repos/mariadb-server-11.8/storage/perfschema/table_events_statements.cc:301
    at /home/dan/repos/mariadb-server-11.8/storage/perfschema/table_events_statements.cc:946
    at /home/dan/repos/mariadb-server-11.8/storage/perfschema/table_events_statements.cc:904
    at /home/dan/repos/mariadb-server-11.8/storage/perfschema/ha_perfschema.cc:356
    at /home/dan/repos/mariadb-server-11.8/sql/handler.cc:3816
    at /home/dan/repos/mariadb-server-11.8/sql/sql_select.cc:24408
    fields=@0x7e6ac96da818: {<base_list> = {<Sql_alloc> = {<No data fields>}, first = 0x7e6ac96dab78, last = 0x7e6ac96daf78, elements = 4}, <No data fields>},
    conds=0x0, og_num=0, order=0x0, group=0x0, having=0x0, proc_param=0x0, select_options=2164525824, result=0x7e6ac96dc0f0, unit=0x7e5ac9694768,
    select_lex=0x7e6ac96da560) at /home/dan/repos/mariadb-server-11.8/sql/sql_select.cc:5426
    at /home/dan/repos/mariadb-server-11.8/sql/sql_select.cc:634
    at /home/dan/repos/mariadb-server-11.8/sql/sql_parse.cc:4014
    rawbuf=0x7e6ac96da438 "SELECT event_name, digest, digest_text, sql_text FROM events_statements_history_long", length=84, parser_state=0x7b9abb8de7b0)
    at /home/dan/repos/mariadb-server-11.8/sql/sql_parse.cc:7953
    at /home/dan/repos/mariadb-server-11.8/sql/sql_parse.cc:1923

1123	    return (uint32) Charset::numchars(ptr(), end());
(rr) p *this
$3 = {<Charset> = {m_charset = 0x77b69a0 <my_charset_latin1>}, <Binary_string> = {<Sql_alloc> = {<No data fields>}, Ptr = 0x0, str_length = 0,
    Alloced_length = 0, extra_alloc = 0, alloced = false, thread_specific = false}, <No data fields>}
(rr) p end()
$4 = 0x0

grooverdan committed:
null protocol

At Protocol::store_string_aux (#6) we have a 0 ptr
and rather than a character conversion we just send it.

From main.plugin_auth_qa test.

Thread 2 hit Breakpoint 1, 0x00000000005c3000 in __ubsan_on_report ()
(rr) bt
    from_length=0, from_cs=0x7989de0 <my_charset_utf8mb4_bin>, errors=0x7ba6041d15a0) at /home/dan/repos/mariadb-server-11.8/sql/sql_string.h:53
    to_cs=0x77b69a0 <my_charset_latin1>) at /home/dan/repos/mariadb-server-11.8/sql/protocol.cc:114
    tocs=0x77b69a0 <my_charset_latin1>) at /home/dan/repos/mariadb-server-11.8/sql/protocol.cc:1461
    tocs=0x77b69a0 <my_charset_latin1>) at /home/dan/repos/mariadb-server-11.8/sql/protocol.cc:1500
    at /home/dan/repos/mariadb-server-11.8/sql/protocol.h:153
    buf=0x7ba60465e860) at /home/dan/repos/mariadb-server-11.8/sql/sql_type.cc:7718
    buf=0x7ba60465e860) at /home/dan/repos/mariadb-server-11.8/sql/sql_type.h:5762
    at /home/dan/repos/mariadb-server-11.8/sql/item.h:1268
    at /home/dan/repos/mariadb-server-11.8/sql/item.cc:8940
    at /home/dan/repos/mariadb-server-11.8/sql/item.cc:9712
    at /home/dan/repos/mariadb-server-11.8/sql/protocol.cc:1359
    items=@0x7e76158da7f8: {<base_list> = {<Sql_alloc> = {<No data fields>}, first = 0x7e76158dab50, last = 0x7e76158dae00, elements = 3}, <No data fields>})
    at /home/dan/repos/mariadb-server-11.8/sql/sql_class.cc:3306
    items=@0x7e76158da7f8: {<base_list> = {<Sql_alloc> = {<No data fields>}, first = 0x7e76158dab50, last = 0x7e76158dae00, elements = 3}, <No data fields>},
    u=0x7e6615894768, sent=0) at /home/dan/repos/mariadb-server-11.8/sql/sql_class.h:6380
    at /home/dan/repos/mariadb-server-11.8/sql/sql_select.cc:25757
    at /home/dan/repos/mariadb-server-11.8/sql/sql_select.cc:24644
    at /home/dan/repos/mariadb-server-11.8/sql/sql_select.cc:24411
    fields=@0x7e76158da7f8: {<base_list> = {<Sql_alloc> = {<No data fields>}, first = 0x7e76158dab50, last = 0x7e76158dae00, elements = 3}, <No data fields>},
    conds=0x7e76158db7a8, og_num=0, order=0x0, group=0x0, having=0x0, proc_param=0x0, select_options=2164525824, result=0x7eb615888100, unit=0x7e6615894768,
    select_lex=0x7e76158da540) at /home/dan/repos/mariadb-server-11.8/sql/sql_select.cc:5426
    at /home/dan/repos/mariadb-server-11.8/sql/sql_select.cc:634
    at /home/dan/repos/mariadb-server-11.8/sql/sql_parse.cc:4014
    rawbuf=0x7e76158da438 "SELECT user,plugin,authentication_string FROM mysql.user WHERE user != 'root'", length=77, parser_state=0x7ba6048ee7b0)
    at /home/dan/repos/mariadb-server-11.8/sql/sql_parse.cc:7953
    at /home/dan/repos/mariadb-server-11.8/sql/sql_parse.cc:1923
